### DISCUSSION TITLE
How to do tensor product expansions with e3nn.

### DISCUSSION THREAD

**Discussion by pfebrer** [e3nn/e3nn#375] [Q&A]
Hi, thanks for developing and publishing the package, it is really nice!

In this paper: https://proceedings.neurips.cc/paper/2021/hash/78f1893678afbeaa90b1fa01b9cfb860-Abstract.html, you define tensor product expansions as going from "direct sum of irreps" to a matrix. We are trying to understand how to do that operation in `e3nn`, but we don't seem to find anything in the documentation.

Is there some functionality exposed for doing this? We are also interested in the special case there is simmetry in the matrix. E.g: going from "1x0e+2x0e" into a 3x3 symmetric matrix. 

We have seen that these kind of things can be done with `CartesianTensor` for 3x3 matrices, but we have found no way of doing it for matrices of arbitrary order, e.g. a 5x5 matrix that is the outer product of l=2 tensors.

Thank you!

**Comment by pfebrer** [comment]
Oh, and we'd also like to do that for matrices that are not square, e,g, a 3x5 matrix representing the tensor product between l=1 and l=2 tensors, if that makes sense.

**Comment by blondegeek** [comment]
In general this transformation is not unique (e.g. in the case where you
have multiple copies of a specific irrep). However, you can find one
possible change of basis by using `o3.ReducedTensorProduct`. For example,
if you want to recover a 3 x 5 matrix where the first dimension spans L=1
and the second dimension spans L=2, then you can use the following:
```
rtp = o3.ReducedTensorProduct(‘ij’, i=‘1o’, j=‘2e’)
rtp.change_of_basis
```
This will return a matrix with shape [15, 3, 5] that you can use to
“convert” your `1o + 2o + 3o` output back to matrix form.

**Reply by pfebrer** [reply]
Thanks for the quick answer!

> In general this transformation is not unique (e.g. in the case where you
> have multiple copies of a specific irrep).

Do you mean that if doing something like this

```python
rtp = o3.ReducedTensorProducts("ij", i="2x1o", j="1o")
```

then `rtp.change_of_basis` is not unique?

Could you point me to some reference for understanding why this transformation is not unique, if there is any?

We are trying to fit a matrix with the same symmetries as the Hamiltonian, so we are trying to understand if, for a given atom, we need to predict block by block or we can generate the full atom-atom interaction at once.

As an example, if you have an atom with a basis set of 1 s orbital and 1 set of p orbitals, its self interaction block is a 4x4 symmetric matrix. Does it make sense to do:

```python
rtp = o3.ReducedTensorProducts("ij=ji", i="0e + 1o")
rtp
```
```
ReducedTensorProducts(
    in: 1x0e+1x1o times 1x0e+1x1o
    out: 2x0e+1x1o+1x2e
)
```

and then conclude from that that if you "predict" a tensor with irreps `2x0e+1x1o+1x2e` you will be able to generate the appropiate 4x4 block with the correct symmetry? It seems reasonable since it is the same you get if you think about it for each combination of irreps individually.

Note: I updated this last part of code because I noticed I had made a mistake :)
